File: app/src/main/java/com/rubyhuntersky/peregrine/ui/SellDialogFragment.java
package com.rubyhuntersky.peregrine.ui;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatDialogFragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.SpinnerAdapter;
import android.widget.TextView;

import com.rubyhuntersky.peregrine.AssetPrice;
import com.rubyhuntersky.peregrine.R;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import rx.Observable;
import rx.functions.Action1;
import rx.functions.Func1;
import rx.subjects.BehaviorSubject;

/**
 * @author wehjin
 * @since 1/19/16.
 */
public class SellDialogFragment extends AppCompatDialogFragment {

    public static final String AMOUNT_KEY = "amountKey";
    public static final String PRICES_KEY = "pricesKey";
    public static final String SELECTED_PRICE_KEY = "selectedPriceKey";
    private BigDecimal amount = BigDecimal.ZERO;
    private BehaviorSubject<List<AssetPrice>> prices = BehaviorSubject.create(Collections.<AssetPrice>emptyList());
    private BehaviorSubject<AssetPrice> selectedPrice = BehaviorSubject.create(new AssetPrice("-", BigDecimal.ZERO));

    public static SellDialogFragment create(BigDecimal amount, List<AssetPrice> priceOptions,
          AssetPrice selectedPrice) {

        final Bundle arguments = new Bundle();
        arguments.putSerializable(AMOUNT_KEY, amount);
        arguments.putParcelableArrayList(PRICES_KEY, new ArrayList<>(priceOptions));
        arguments.putParcelable(SELECTED_PRICE_KEY, selectedPrice);

        final SellDialogFragment sellDialogFragment = new SellDialogFragment();
        sellDialogFragment.setArguments(arguments);
        return sellDialogFragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(STYLE_NO_TITLE, 0);

        amount = (BigDecimal) getArguments().getSerializable(AMOUNT_KEY);

        final Bundle bundle = savedInstanceState == null ? getArguments() : savedInstanceState;
        prices.onNext(bundle.<AssetPrice>getParcelableArrayList(PRICES_KEY));
        selectedPrice.onNext(bundle.<AssetPrice>getParcelable(SELECTED_PRICE_KEY));
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        outState.putParcelableArrayList(PRICES_KEY, new ArrayList<>(prices.getValue()));
        outState.putParcelable(SELECTED_PRICE_KEY, selectedPrice.getValue());
        super.onSaveInstanceState(outState);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.setCanceledOnTouchOutside(true);
        return dialog;
    }


    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
          @Nullable Bundle savedInstanceState) {
        final View inflate = inflater.inflate(R.layout.fragment_sell, container, false);

        final TextView amountText = (TextView) inflate.findViewById(R.id.amountText);
        amountText.setText("Sell " + UiHelper.getCurrencyDisplayString(amount));

        final Spinner priceSpinner = (Spinner) inflate.findViewById(R.id.priceSpinner);
        priceSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                selectedPrice.onNext(prices.getValue().get(position));
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                // Do nothing
            }
        });
        prices.map(new Func1<List<AssetPrice>, List<String>>() {
            @Override
            public List<String> call(List<AssetPrice> prices) {
                return Observable.from(prices).map(new Func1<AssetPrice, String>() {
                    @Override
                    public String call(AssetPrice price) {
                        return price.name + " " + UiHelper.getCurrencyDisplayString(price.amount);
                    }
                }).toList().toBlocking().first();
            }
        }).map(new Func1<List<String>, SpinnerAdapter>() {
            @Override
            public SpinnerAdapter call(List<String> strings) {
                final ArrayAdapter<String> adapter = new ArrayAdapter<>(getActivity(),
                      android.R.layout.simple_spinner_item, strings);
                adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                return adapter;
            }
        }).subscribe(new Action1<SpinnerAdapter>() {
            @Override
            public void call(SpinnerAdapter spinnerAdapter) {
                priceSpinner.setAdapter(spinnerAdapter);
                priceSpinner.setSelection(prices.getValue().indexOf(selectedPrice.getValue()));
            }
        });

        final TextView sharesText = (TextView) inflate.findViewById(R.id.sharesText);
        selectedPrice.subscribe(new Action1<AssetPrice>() {
            @Override
            public void call(AssetPrice price) {
                sharesText.setText(price.getSharesString(amount));
            }
        });
        return inflate;
    }


}

File: app/src/main/java/com/rubyhuntersky/peregrine/ui/SellDialogFragment.kt
package com.rubyhuntersky.peregrine.ui

import android.os.Bundle
import android.support.design.widget.BottomSheetDialogFragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import com.rubyhuntersky.peregrine.R
import com.rubyhuntersky.peregrine.model.AssetNamePrice
import com.rubyhuntersky.peregrine.model.GroupSaleOption
import com.rubyhuntersky.peregrine.utility.toSharesDisplayString
import com.rubyhuntersky.peregrine.utility.withArguments
import kotlinx.android.synthetic.main.cell_account_name_shares.view.*
import kotlinx.android.synthetic.main.fragment_sell.*
import rx.Subscription
import rx.subjects.BehaviorSubject
import java.math.BigDecimal
import java.util.*

class SellDialogFragment : BottomSheetDialogFragment() {

    val TAG = SellDialogFragment::javaClass.name

    private val amount by lazy { arguments!!.getSerializable(AMOUNT_KEY) as BigDecimal }
    private val groupSaleOptions by lazy { arguments!!.getParcelableArrayList<GroupSaleOption>(OPTIONS_KEY) }
    private val groupAssetPrices get() = groupSaleOptions.map { AssetNamePrice(it.assetName, it.assetPrice) }
    private val viewModel by lazy { PriceSelectionViewModel(view!!) }

    private val indexSubject = BehaviorSubject.create(0)
    private var indexSubscription: Subscription? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val data = savedInstanceState ?: arguments!!
        indexSubject.onNext(data.getInt(SELECTED_INDEX_KEY))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putInt(SELECTED_INDEX_KEY, indexSubject.value)
        super.onSaveInstanceState(outState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        return inflater.inflate(R.layout.fragment_sell, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        indexSubject.subscribe { index ->
            val model = PriceSelectionModel(amount, groupAssetPrices, groupAssetPrices[index], "Sell")
            viewModel.bind(model, { _, _ -> })

            val groupSaleOption = groupSaleOptions[index]
            val accountSaleOptions = groupSaleOption.accountSaleOptions
            Log.i(TAG, "AccountSaleOptions: $accountSaleOptions")
            accounts_list.removeAllViews()
            accountSaleOptions.forEach {
                Log.i(TAG, "AccountSaleOption: $it")
                val cell = View.inflate(context, R.layout.cell_account_name_shares, null)
                cell.account_name.text = it.accountTitle
                cell.account_shares.text = it.sharesAvailable.toSharesDisplayString()
                accounts_list.addView(cell, MATCH_PARENT, WRAP_CONTENT)
            }
        }
    }

    override fun onDestroyView() {
        indexSubscription?.unsubscribe()
        super.onDestroyView()
    }


    companion object {

        private val AMOUNT_KEY = "amountKey"
        private val OPTIONS_KEY = "optionsKey"
        private val SELECTED_INDEX_KEY = "selectedOptionIndexKey"

        @JvmStatic
        fun create(cashGoal: BigDecimal,
                   options: List<GroupSaleOption>,
                   initialOptionIndex: Int): SellDialogFragment {

            return SellDialogFragment().withArguments {
                putSerializable(AMOUNT_KEY, cashGoal)
                putParcelableArrayList(OPTIONS_KEY, ArrayList(options))
                putInt(SELECTED_INDEX_KEY, initialOptionIndex)
            }
        }
    }
}
